src/sentry/api/endpoints/user_authenticator_enroll.py

```diff
@@ -0,0 +1,202 @@
+from __future__ import absolute_import
+
+from rest_framework import serializers, status
+from rest_framework.permissions import IsAuthenticated
+from rest_framework.response import Response
+
+import petname
+
+from sentry.api.bases.user import UserEndpoint, UserPermission
+from sentry.api.decorators import sudo_required
+from sentry.api.serializers import serialize
+from sentry.models import Authenticator
+from sentry.security import capture_security_activity
+
+
+class TotpRestSerializer(serializers.Serializer):
+    otp = serializers.CharField(
+        label='One-time password',
+        help_text='Code from authenticator',
+        required=True,
+        max_length=20
+    )
+
+
+class SmsRestSerializer(serializers.Serializer):
+    phone = serializers.CharField(
+        label="Phone number",
+        help_text="Phone number to send SMS code",
+        required=True,
+        max_length=20,
+    )
+    otp = serializers.CharField(
+        label='One-time password',
+        help_text='Code from authenticator',
+        required=False,
+        max_length=20
+    )
+
+
+class U2fRestSerializer(serializers.Serializer):
+    deviceName = serializers.CharField(
+        label='Device name',
+        required=False,
+        max_length=60,
+        default=lambda: petname.Generate(2, ' ', letters=10).title(),
+    )
+
+serializer_map = {
+    'totp': TotpRestSerializer,
+    'sms': SmsRestSerializer,
+    'u2f': U2fRestSerializer,
+}
+
+
+def get_serializer_field_metadata(serializer, fields=None):
```

Comment:
Sooo, this takes the rest framework serializer and serializes its fields to send metadata to the frontend in order to render. This bloats the API endpoint since we're always sending the form fields, but can potentially simplify the frontend code a bit. This is kind of experimental since these endpoints have very small forms.

If we were to get rid of this, how would we go about solving backend generated initial data (e.g.`deviceName`)?

Reply:
we do something kinda similar with plugins, so sgtm